lib/_http_client.js

```diff
@@ -87,6 +87,14 @@ function ClientRequest(options, cb) {
     }
     if (host && !this.getHeader('host') && setHost) {
       var hostHeader = host;
+
+      // For the Host header, ensure that IPv6 addresses are enclosed
+      // in square brackets, as defined by URI formatting
+      // https://tools.ietf.org/html/rfc3986#section-3.2.2
+      if (net.isIPv6(hostHeader)) {
```

Comment:
Host value doesn't pass through proper validation. It actually relies on either `url.parse()` to produce a host value, or simply takes `options.hostname` || `options.host`. At this stage, especially if `options` object is passed to the `request()`, `hostHeader` could even be `www.domain.tld:3000`, before reaching my code; (_which is not wrong, only unsupported atm_). In this case, a simple `:` check would yield an invalid Host header `[www.domain.tld:3000]`.

Secondly, the RFC says that `if (IP is IPv6) then [IP]`, which in my view makes it mandatory to make absolutely sure that whatever I enclose within square brackets, is an IPv6 address and not anything else (even if invalid).

I've been considering all of the above right from my initial PR, including the fact that `indexOf` is faster than `isIPv6`. However, since all points of view express advantages and disadvantages, I also opted for `net.isIPv6` in the end, because that allows my fix to avoid type assumptions and ultimately make the code terse, and enforces the RFC statement.

While the above explains my preference for `net.isIPv6`, if there's consensus on reverting it back to `:` matching, I'll happily do it.